File: rtweekend_lib/hittable_list.go
package rtweekendlib

import (
	"container/list"
)

type HittableList struct {
	objects list.List
}

func (hl *HittableList) Clear() {
	hl.objects = *new(list.List)
}

func (hl *HittableList) Add(object Hittable) {
	hl.objects.PushBack(object)
}

func (hl *HittableList) Hit(r *Ray, t_min, t_max float64, rec *HitRecord) bool {
	tempRec := new(HitRecord)
	hitAnything := false
	closestSoFar := t_max

	for object := hl.objects.Front(); object != nil; object = object.Next() {
		if object.Value.(Hittable).Hit(r, t_min, closestSoFar, tempRec) {
			hitAnything = true
			closestSoFar = tempRec.T()
			rec = tempRec
		}
	}

	return hitAnything
}

func NewHittableList(object Hittable) *HittableList {
	hl := new(HittableList)

	hl.Add(object)

	return hl
}

File: include/hittable_list/hittable_list.c
#include "hittable_list.h"

#include "../sphere/sphere.h"

bool hit_HittableList(
        const HittableList *hl,
        const Ray *r,
        double t_min, double t_max,
        HitRecord *rec
) {
    HitRecord tmp_rec;
    bool hit_anything = false;
    double closest_so_far;

    // Loop through objects in HittableList here!

    return hit_anything;
}
